core.js

```diff
@@ -60,16 +44,120 @@ function _check(buffer, headers, options) {
 }
 
 export async function fileTypeFromTokenizer(tokenizer) {
-	try {
-		return new FileTypeParser().parse(tokenizer);
-	} catch (error) {
-		if (!(error instanceof strtok3.EndOfStreamError)) {
-			throw error;
-		}
+	return new FileTypeParser().fromTokenizer(tokenizer);
+}
+
+export class TokenizerPositionError extends Error {
+	constructor(message = 'Tokenizer position changed without detecting a file type.') {
+		super(message);
+		this.name = 'TokenizerPositionError';
 	}
 }
 
-class FileTypeParser {
+export class FileTypeParser {
+	constructor(options) {
+		this.detectors = options?.customDetectors;
+
+		this.runCustomDetectors = this.#runCustomDetectors.bind(this);
+		this.fromTokenizer = this.fromTokenizer.bind(this);
+		this.fromBuffer = this.fromBuffer.bind(this);
+		this.parse = this.parse.bind(this);
+	}
+
+	async #runCustomDetectors(tokenizer) {
+		const {detectors} = this;
+		const initialPosition = tokenizer?.position;
+
+		for (const detector of detectors || []) {
+			const fileType = await detector(tokenizer);
+			if (fileType) {
+				return fileType;
+			}
+
+			if (initialPosition !== tokenizer?.position) {
+				throw new TokenizerPositionError();
+			}
+		}
+
+		return undefined;
+	}
+
+	async fromTokenizer(tokenizer) {
+		const {parse, runCustomDetectors} = this;
+		try {
+			const customFileType = await runCustomDetectors(tokenizer);
+			return (customFileType || await parse(tokenizer));
+		} catch (error) {
+			if (!(error instanceof strtok3.EndOfStreamError) && !(error instanceof TokenizerPositionError)) {
+				throw error;
+			}
+		}
+	}
+
+	async fromBuffer(input) {
+		if (!(input instanceof Uint8Array || input instanceof ArrayBuffer)) {
+			throw new TypeError(`Expected the \`input\` argument to be of type \`Uint8Array\` or \`Buffer\` or \`ArrayBuffer\`, got \`${typeof input}\``);
+		}
+
+		const buffer = input instanceof Uint8Array ? input : new Uint8Array(input);
+
+		if (!(buffer?.length > 1)) {
+			return;
+		}
+
+		return this.fromTokenizer(strtok3.fromBuffer(buffer));
+	}
+
+	async fromBlob(blob) {
+		const buffer = await blob.arrayBuffer();
+		return this.fromBuffer(new Uint8Array(buffer));
+	}
+
+	async fromStream(stream) {
+		const tokenizer = await strtok3.fromStream(stream);
+		try {
+			return await this.fromTokenizer(tokenizer);
+		} finally {
+			await tokenizer.close();
+		}
+	}
+
+	async stream(readableStream, options = {}) {
```

Comment:
Suggestion, maybe good to articulate the different functionality here.
```suggestion
	async toDetectingStream(readableStream, options = {}) {
```

Reply:
Done.